metaseq/distributed/utils.py

```diff
@@ -200,6 +205,43 @@ def distributed_main(i, main, cfg: MetaseqConfig, kwargs):
         torch.distributed.barrier(get_global_group())
 
 
+def _spawn_helper(main, cfg, kwargs):
+    """
+    Perform a fork() to many processes.
+
+    Intentionally runs the rank0 in the main process so that signals
+    can be more easily caught and we can cleanup processes.
+    """
+    # Launch multiple subprocesses
+    spawncontext = torch.multiprocessing.start_processes(
+        distributed_main,
+        # need to give rank offset as 1 to cover the fact that the main
+        # process is rank 0, but that spawn() doesn't let you control rank
+        (main, cfg, kwargs),
+        nprocs=min(
+            torch.cuda.device_count(),
+            cfg.distributed_training.distributed_world_size,
+        ),
+        join=False,
+        start_method='spawn',
+    )
+
+    try:
+        # -1 because we offset by +1 inside distributed_main when using
+        # spawn_helper
+        retval = distributed_main(-1, main, cfg, kwargs)
```

Comment:
Main motivation behind getting rid of `distributed_no_spawn` gate is to simplify things like this.

Reply:
This would still be the case. The 0-indexing comes from torch.multiprocessing.spawn.